File: cxsever/www/js/lib/lineardoc/Normalizer.js
import sax from 'sax';
import { get_close_tag_html, get_open_tag_html } from './utils.js';

/**
 * Escape text for inclusion in HTML, not inside a tag.
 *
 * @private
 * @param {string} str String to escape
 * @return {string} Escaped version of the string
 */
function esc(str) {
	return str.replace(/[&<>]/g, (ch) => '&#' + ch.charCodeAt(0) + ';');
}

/**
 * Parser to normalize XML.
 *
 * @class
 * @constructor
 */
class Normalizer extends sax.SAXParser {
	constructor() {
		super(false, {
			lowercase: true
		});
	}

	init() {
		this.doc = [];
		this.tags = [];
	}

	onopentag(tag) {
		this.tags.push(tag);
		this.doc.push(get_open_tag_html(tag));
	}

	onclosetag(tagName) {
		const tag = this.tags.pop();
		if (tag.name !== tagName) {
			throw new Error('Unmatched tags: ' + tag.name + ' !== ' + tagName);
		}
		this.doc.push(get_close_tag_html(tag));
	}

	ontext(text) {
		this.doc.push(esc(text));
	}

	getHtml() {
		return this.doc.join('');
	}

}

export default Normalizer;

File: cxsever/www/python/lib/lineardoc/normalizer.py
"""
Normalizer - Parser to normalize XML.
"""

from lxml import etree

from . import utils as Utils


def esc(s):
    """Escape text for inclusion in HTML."""
    return s.replace("&", "&#38;").replace("<", "&#60;").replace(">", "&#62;")


class Normalizer:
    """Parser to normalize XML."""

    def __init__(self):
        """Initialize the normalizer."""
        self.lowercase = True

    def init(self):
        """Initialize state for parsing."""
        self.doc = []
        self.tags = []

    def write(self, html):
        """
        Parse and normalize HTML.

        Args:
            html: HTML string to normalize
        """
        parser = etree.HTMLParser()
        try:
            tree = etree.fromstring(html, parser)
            self._process_element(tree)
        except Exception:
            # Try with wrapping
            try:
                tree = etree.fromstring(f"<div>{html}</div>", parser)
                for child in tree:
                    self._process_element(child)
            except Exception as e:
                raise Exception(f"Failed to parse HTML: {e}")

    def _process_element(self, element):
        """Process an element recursively."""
        tag_name = element.tag.lower() if self.lowercase else element.tag

        # Create tag dict
        tag = {"name": tag_name, "attributes": dict(element.attrib)}

        self.on_open_tag(tag)

        # Process text content
        if element.text:
            self.on_text(element.text)

        # Process children
        for child in element:
            self._process_element(child)
            # Process tail text after child
            if child.tail:
                self.on_text(child.tail)

        self.on_close_tag(tag_name)

    def on_open_tag(self, tag):
        """Handle open tag event."""
        self.tags.append(tag)
        self.doc.append(Utils.get_open_tag_html(tag))

    def on_close_tag(self, tag_name):
        """Handle close tag event."""
        tag = self.tags.pop()
        if tag["name"] != tag_name:
            raise Exception(f'Unmatched tags: {tag["name"]} !== {tag_name}')
        self.doc.append(Utils.get_close_tag_html(tag))

    def on_text(self, text):
        """Handle text event."""
        self.doc.append(esc(text))

    def get_html(self):
        """
        Get the normalized HTML.

        Returns:
            Normalized HTML string
        """
        return "".join(self.doc)
